rfc/rfc-63/rfc-63.md

```diff
@@ -0,0 +1,370 @@
+  contributor license agreements.  See the NOTICE file distributed with
+  this work for additional information regarding copyright ownership.
+  The ASF licenses this file to You under the Apache License, Version 2.0
+  (the "License"); you may not use this file except in compliance with
+  the License.  You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-->
+
+# RFC-63: Index Function for Optimizing Query Performance
+
+## Proposers
+
+- @yihua
+- @alexeykudinkin
+
+## Approvers
+
+- @vinothchandar
+- @xushiyan
+- @nsivabalan
+
+## Status
+
+JIRA: [HUDI-512](https://issues.apache.org/jira/browse/HUDI-512)
+
+## Abstract
+
+In this RFC, we address the problem of accelerating queries containing predicates based on functions defined on a
+column, by introducing **Index Function**, a new indexing capability for efficient file pruning.
+
+## Background
+
+To make the queries finish faster, one major optimization technique is to scan less data by pruning rows that are not
+needed by the query. This is usually done in two ways:
+
+- **Partition pruning**:  The partition pruning relies on a table with physical partitioning, such as Hive partitioning.
+  A partitioned table uses a chosen column such as the date of `timestamp` and stores the rows with the same date to the
+  files under the same folder or physical partition, such as `date=2022-10-01/`. When the predicate in a query
+  references the partition column of the physical partitioning, the files in the partitions not matching the predicate
+  are filtered out, without scanning. For example, for the predicate `date between '2022-10-01' and '2022-10-02'`, the
+  partition pruning only returns the files from two partitions, `2022-10-01` and `2022-10-02`, for further processing.
+  The granularity of the pruning is at the partition level.
+
+
+- **File pruning**:  The file pruning carries out the pruning of the data at the file level, with the help of file-level
+  or record-level index. For example, with column stats index containing minimum and maximum values of a column for each
+  file, the files falling out of the range of the values compared to the predicate can be pruned. For a predicate
+  with `age < 20`, the file pruning filters out a file with columns stats of `[30, 40]` as the minimum and maximum
+  values of the column `age`.
+
+While Apache Hudi already supports partition pruning and file pruning with data skipping for different query engines, we
+recognize that the following use cases need better query performance and usability:
+
+- File pruning based on functions defined on a column
```

Comment:
what functions do we going to support? years/months/days/hours defined in spark?

Reply:
yes some date and string functions that preserve ordering